### PR TITLE
Discussion of JSON output

### PR DESCRIPTION
**Pull request by Honny1**
The purpose of this PR is to agree on the content of the JSON output. So feel free to discuss.

PR includes:
* Example of JSON output that contains all information that `openscap-report` can offer (`JSON-EVERYTHING`)
* Schema of `JSON-EVERYTHING`
* Comparison of parsed content between `openscap_parser` and `openscap-report` (I plan to add the missing information to the `openscap-report` later.)
* Example of JSON output
* Schema of `JSON`
 

 [JSON schema standard](https://json-schema.org/)

### CODE REVIEW

**Comment by jan-cerny** [comment]
Maybe the references with the same "href" value should be merged to a single item that would contain a list of identifiers? currently there is a separate item for each identifier which causes the duplication of the "href" value.

**Comment by evgenyz** [comment]
Can we please have a smaller sample report? 600K lines is way over the board for an example.

**Comment by Honny1** [comment]
@evgenyz I'm adding a small JSON example, but many fields are `null`.  [Used ARF file](https://github.com/OpenSCAP/openscap-report/blob/master/tests/test_data/arf_simple_rule_pass.xml)

**Comment by Honny1** [comment]
I would like to follow up discussion about a `null` value in PR #105. Which is the best value to use for cases when there is no data?

Here are the options:
* Use the `null` value. 
  * ++ 
    * It is obvious there is nothing.
    * It is the same for every data type.
  * --
    * Value needs to be checked before usage. 
    * Iteration over null is not possible.
    * Problematic implementation can cause many errors. 
 * Use empty values as an empty array, empty string, etc.
   * ++
     * Value doesn't need to be checked before usage.
     * Iteration is possible.
     * Better implementation.
   * --
     * Every data type have own different empty value.
     * For example, the OVAL tree object hasn't got any empty value. So back to `null` again?
* Don't put fields that have empty values or `null` to JSON.
  * ++
    * Saves space.
  * --
    *  Can cause errors when the key is not present in JSON.
* Combination of the first and second options.
  * Don't be strict about not using the `null` value. Use null value when there is no empty option.
    * ++
      * Not every value doesn't need to be checked before usage.
    * --
      * It is not the same for every data type.

**Comment by Honny1** [comment]
I decided to use empty arrays and an empty dictionary for keys that contains an array or dictionary. Other keys will be null or optional.

**Comment by Honny1** [comment]
IDEA: 
Create several versions of JSON that will extend default JSON. 
For example, JSON will be extended with: 
- OVAL definitions 
- OVAL and CPE trees 
- Groups

File: JSON/very_small_example_report.json [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+			"title": "This rule always passes",
+			"result": "pass",
+			"multi_check": false,
+			"time": "2021-11-30T12:27:51+01:00",
+			"severity": "Unknown",
+			"identifiers": [],
+			"references": [],
+			"description": "Description",
+			"rationale": "",
+			"warnings": [],
```
Comment (marcusburghardt) [inline JSON/very_small_example_report.json:30]: There are different types of warning in the content, like `general`, `performance`, `functionality` and `regulatory`. And the types can be repeated in a rule. So, the json output should identify the `type` and the message itself. Could you extend this?

File: JSON/very_small_example_report.json [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+			"platforms": [],
+			"oval_definition_id": "oval:x:def:1",
+			"messages": [],
+			"remediations": [],
```
Comment (marcusburghardt) [inline JSON/very_small_example_report.json:34]: How would be a remediation represented here? Would be the bash script or the Ansible playbook, for example? Or this key would only inform the available "flavors" for that rule?

File: JSON/very_small_example_report.json [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+			"oval_tree": {
```
Comment (marcusburghardt) [inline JSON/very_small_example_report.json:35]: This three could became complex. I suggest to include some complex rules in the tests. For example:
- https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_faillock_deny/oval/shared.xml

This rule contains multiple tests and variables.

File: JSON/very_small_example_report.json (hunk 2 above)
Comment (marcusburghardt) [reply JSON/very_small_example_report.json:34]: Never mind. I just saw in the schema, so I will comment there.

File: JSON/JSON-EVERYTHING_schema.json [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                    "title": "Disruption",
+                    "default": "",
+                    "type": "string"
+                },
+                "strategy": {
+                    "title": "Strategy",
+                    "default": "",
+                    "type": "string"
+                },
+                "fix": {
```
Comment (marcusburghardt) [inline JSON/JSON-EVERYTHING_schema.json:171]: I am not familiar to the JSON standard in details and didn't have opportunity to take a look in the openscap-report, but I believe an remediation which potentially contains json in its content, by any reason, would not cause problems here, right? I don't have an example of this happening in the content, but just exercising the threat modeling here.

File: JSON/JSON-EVERYTHING_schema.json [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+                    "title": "Description",
+                    "default": "",
+                    "type": "string"
+                },
+                "rationale": {
+                    "title": "Rationale",
+                    "default": "",
+                    "type": "string"
+                },
+                "warnings": {
```
Comment (marcusburghardt) [inline JSON/JSON-EVERYTHING_schema.json:366]: Maybe the warning should be grouped by type?

File: JSON/very_small_example_report.json (hunk 1 above)
Comment (Honny1) [reply JSON/very_small_example_report.json:30]: Yes, I can extend this.

File: JSON/JSON-EVERYTHING_schema.json (hunk 4 above)
Comment (Honny1) [reply JSON/JSON-EVERYTHING_schema.json:171]: It doesn't cause a problem. The content of the fix is escaped in this context.

File: JSON/very_small_example_report.json (hunk 3 above)
Comment (Honny1) [reply JSON/very_small_example_report.json:35]: I added an example rule [xccdf_org.ssgproject.content_rule_accounts_passwords_pam_faillock_deny](https://github.com/OpenSCAP/openscap-report/blob/0b329c334df63ec53fcaffc028743e8b63fe7ace/JSON/small_example_with_more_complex_rule.json).

**Comment by Honny1** [comment]
Hello, I have added an example and a schema of the default JSON version. The implementation of this is in PR #130.

**Comment by Honny1** [comment]
I'm closing this PR for inactivity.

**Pull request closed without merging**